geps/gep-1767/index.md

```diff
@@ -0,0 +1,701 @@
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: https://foo.example
+    //
+    // Support: Extended
+    // +listType=set
+    // +kubebuilder:validation:MaxItems=64
+    AllowOrigins []string `json:"allowOrigins,omitempty"`
+ 
+    // AllowCredentials indicates whether the actual cross-origin request 
+    // allows to include credentials.
+    //
+    // The only valid value for the header `Access-Control-Allow-Credentials` is true (case-sensitive).
+    //
+    // The default value of the config AllowCredentials is false.
+    // If the credentials are not allowed in cross-origin requests, 
+    // the gateway will omit the header `Access-Control-Allow-Credentials` entirely.
+    //
+    // Input:
+    //   Origin: https://foo.example
+    //
+    // Config:
+    //   allowCredentials: true
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: https://foo.example
+    //   Access-Control-Allow-Credentials: true
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +kubebuilder:default=false
+    AllowCredentials bool `json:"allowCredentials,omitempty"`
+
+    // AllowMethods indicates which HTTP methods are supported 
+    // for accessing the requested resource.
+    // The method is case-sensitive.
+    //
+    // Config:
+    //   allowMethods: ["GET, PUT, POST, DELETE, PATCH, OPTIONS"]
+    // 
+    // Output:
+    //   Access-Control-Allow-Methods: GET, PUT, POST, DELETE, PATCH, OPTIONS
+    //
+    // A wildcard indicates that the requests with all HTTP methods are allowed.
+    // When responding to a credentialed requests, the gateway must specify 
+    // one or more HTTP methods in the value of the Access-Control-Allow-Methods response header, 
+    // instead of specifying the * wildcard.
+    //
+    // Config:
+    //   allowMethods: ["*"]
+    // 
+    // Output:
+    //   Access-Control-Allow-Methods: ["*"]
+    //
+    // Support: Extended
+    //
+    // +optional
+    // +listType=set
+    // +kubebuilder:validation:MaxItems=16
+    AllowMethods []string `json:"allowMethods,omitempty"`
+ 
+    // AllowHeaders indicates which HTTP request headers are supported 
+    // for accessing the requested resource.
+    // Header names are not case sensitive.
+    //
+    // Config:
+    //   allowHeaders: ["DNT", "Keep-Alive", "User-Agent", "X-Requested-With", "If-Modified-Since", "Cache-Control", "Content-Type", "Range", "Authorization"]
+    // Output:
+    //   Access-Control-Allow-Headers: DNT, Keep-Alive, User-Agent, X-Requested-With, If-Modified-Since, Cache-Control, Content-Type, Range, Authorization
+    //
+    // A wildcard indicates that the requests with all HTTP headers are allowed.
+    // When responding to a credentialed requests, the gateway must specify 
+    // one or more HTTP headers in the value of the Access-Control-Allow-Headers response header, 
+    // instead of specifying the * wildcard.
```

Comment:
I think you may be mixing up the user HTTPRoute config and the server's expected response. a user can legitimately set `*` in the config certainly -- this seems to be saying they cannot set it in the response header.

However, I am not sure why? The spec does not seem to reject `*`? https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers#syntax

Reply:
We had someone in ingress-nginx, request that we updated CORS to accept * and `null` 

https://github.com/kubernetes/ingress-nginx/pull/12402

Reply:
Note that Envoy can't return `*` directly because it treats `*` specially:
https://github.com/envoyproxy/envoy/blob/6445d0dcdd33413e47454221fa80b3cfabea7d4e/source/extensions/filters/http/cors/cors_filter.cc#L159-L163.

Reply:
> Note that Envoy can't return `*` directly because it treats `*` specially: https://github.com/envoyproxy/envoy/blob/6445d0dcdd33413e47454221fa80b3cfabea7d4e/source/extensions/filters/http/cors/cors_filter.cc#L159-L163.

Most gateways support the `*` wildcard as the value of a HTTP header.
Moreover, `*` wildcard is a valid character of a HTTP header based on [rfc7230](https://www.rfc-editor.org/rfc/rfc7230#section-3.2).

Hence, the envoy should support it in future.

Reply:
This field (and this spec in general) is pretty unclear as to which "user" we're talking about.

I suggest using different names for the HTTPRoute owner (maybe "user" is okay here), and the _client_, (should probably use that word instead).

So, I think what @robscott was saying is "What happens if the HTTPRoute owner has configured `*` here?" 

The text says:
```go
    // When responding to a credentialed requests, the gateway must specify 
    // one or more HTTP headers in the value of the Access-Control-Allow-Headers response header, 
    // instead of specifying the * wildcard.
```

What headers should the implementation respond with? Below, @spacewander suggests responding with whatever was sent in the `Access-Control-Request-Headers` field in the OPTIONS request. Whatever we want to do MUST be defined in this field, as what to do in the other cases:
* `*` is configured in this field, and `Access-Control-Request-Headers` is not sent
* One or more values are configured in this field and `Access-Control-Request-Headers` **is** sent
* One or more values are configured in this field and `Access-Control-Request-Headers` **is not** sent

What we are aiming for is that someone who reads the spec should be able to understand enough to write conformance tests for the spec, ideally without looking at any other documents.